modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -3782,12 +3783,12 @@ def _groupby_shuffle(
                 + "https://github.com/modin-project/modin/issues/5926"
             )
 
-        # So this check works only if we have dtypes cache materialized, otherwise the exception will be thrown
-        # inside the kernel and so it will be uncatchable. TODO: figure out a better way to handle this.
-        if self._modin_frame._dtypes is not None and any(
-            isinstance(dtype, pandas.CategoricalDtype)
-            for dtype in self.dtypes[by].values
-        ):
+        # This check materializes dtypes for 'by' columns
+        if isinstance(self._modin_frame._dtypes, ModinDtypes):
+            by_dtypes = self._modin_frame._dtypes.lazy_get(by).get()
+        else:
+            by_dtypes = self.dtypes[by]
+        if any(isinstance(dtype, pandas.CategoricalDtype) for dtype in by_dtypes):
```

Comment:
We're now materializing 'by' dtypes to consistently catch unsupported cases and fallback to an older implementation. Previously, if the dtypes weren't materialized we were raising an exception in the kernel which caused groupby to fail. Since we're now moving this implementation out of experimental mode, we want more stability here in terms of falling back to an implementation that has more coverage